File: python/jsonatapy/_cli/resolve.py
"""Determines the expression source and input source from parsed CLI
arguments. Single source of truth for how positional arguments,
--from-file, and --null-input interact -- mirrors
src/bin/jsonata/resolve.rs::resolve() in the Rust CLI exactly, so both
implementations agree on every case in study/cli_fixtures.json.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ExpressionInline:
    """The expression text came directly from a positional argument."""

    text: str


@dataclass(frozen=True)
class ExpressionFile:
    """The expression text should be read from this file path (--from-file)."""

    path: str


ExpressionSource = ExpressionInline | ExpressionFile


@dataclass(frozen=True)
class InputStdin:
    """Input JSON should be read from stdin."""


@dataclass(frozen=True)
class InputFile:
    """Input JSON should be read from this file path."""

    path: str


@dataclass(frozen=True)
class InputNull:
    """No input is read; the evaluation context is null (see this plan's
    Global Constraints for why this is null, not JSONata Undefined)."""


InputSource = InputStdin | InputFile | InputNull


class ResolveError(Exception):
    """Raised when CLI arguments cannot be resolved into an expression/input source."""


def resolve(
    from_file: str | None,
    positional1: str | None,
    positional2: str | None,
    null_input: bool,
) -> tuple[ExpressionSource, InputSource]:
    """Resolves the expression source and input source from parsed CLI args."""
    data_file: str | None
    expr_source: ExpressionSource

    if from_file is not None:
        if positional2 is not None:
            raise ResolveError(
                "with --from-file, only one positional argument (the input file) is allowed"
            )
        expr_source = ExpressionFile(from_file)
        data_file = positional1
    elif positional1 is not None:
        expr_source = ExpressionInline(positional1)
        data_file = positional2
    else:
        raise ResolveError("missing required argument: EXPRESSION (or use --from-file)")

    if null_input:
        if data_file is not None:
            raise ResolveError("--null-input cannot be combined with an input file argument")
        return expr_source, InputNull()

    if data_file is not None:
        return expr_source, InputFile(data_file)
    return expr_source, InputStdin()

File: src/bin/jsonata/resolve.rs
/// Where the JSONata expression text comes from.
#[derive(Debug, PartialEq)]
pub enum ExpressionSource {
    Inline(String),
    File(String),
}

/// Where the input JSON document comes from.
#[derive(Debug, PartialEq)]
pub enum InputSource {
    Stdin,
    File(String),
    Null,
}

/// Resolves the expression source and input source from parsed CLI
/// arguments. `--from-file` shifts what a bare positional argument means
/// (it becomes the input data file, not the expression); `--null-input`
/// suppresses reading input entirely and conflicts with supplying a data
/// file. See Task 4 of the Phase 1 plan for the full truth table this
/// implements.
pub fn resolve(cli: &crate::Cli) -> Result<(ExpressionSource, InputSource), String> {
    let (expr_source, data_file) = match (&cli.from_file, &cli.positional1, &cli.positional2) {
        (Some(expr_file), data_file, None) => {
            (ExpressionSource::File(expr_file.clone()), data_file.clone())
        }
        (Some(_), _, Some(_)) => {
            return Err(
                "with --from-file, only one positional argument (the input file) is allowed"
                    .to_string(),
            );
        }
        (None, Some(expr), data_file) => {
            (ExpressionSource::Inline(expr.clone()), data_file.clone())
        }
        (None, None, _) => {
            return Err("missing required argument: EXPRESSION (or use --from-file)".to_string());
        }
    };

    if cli.null_input {
        if data_file.is_some() {
            return Err("--null-input cannot be combined with an input file argument".to_string());
        }
        Ok((expr_source, InputSource::Null))
    } else {
        match data_file {
            Some(f) => Ok((expr_source, InputSource::File(f))),
            None => Ok((expr_source, InputSource::Stdin)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::Cli;

    fn cli(
        from_file: Option<&str>,
        positional1: Option<&str>,
        positional2: Option<&str>,
        null_input: bool,
    ) -> Cli {
        Cli {
            compact: false,
            raw_output: false,
            null_input,
            from_file: from_file.map(String::from),
            arg: Vec::new(),
            argjson: Vec::new(),
            positional1: positional1.map(String::from),
            positional2: positional2.map(String::from),
        }
    }

    #[test]
    fn plain_expression_and_stdin() {
        let c = cli(None, Some("name"), None, false);
        assert_eq!(
            resolve(&c),
            Ok((ExpressionSource::Inline("name".into()), InputSource::Stdin))
        );
    }

    #[test]
    fn plain_expression_and_file() {
        let c = cli(None, Some("name"), Some("data.json"), false);
        assert_eq!(
            resolve(&c),
            Ok((
                ExpressionSource::Inline("name".into()),
                InputSource::File("data.json".into())
            ))
        );
    }

    #[test]
    fn missing_expression_is_an_error() {
        let c = cli(None, None, None, false);
        assert!(resolve(&c).is_err());
    }

    #[test]
    fn null_input_with_no_data_file_is_null_source() {
        let c = cli(None, Some("$now()"), None, true);
        assert_eq!(
            resolve(&c),
            Ok((ExpressionSource::Inline("$now()".into()), InputSource::Null))
        );
    }

    #[test]
    fn null_input_with_data_file_is_an_error() {
        let c = cli(None, Some("name"), Some("data.json"), true);
        assert!(resolve(&c).is_err());
    }

    #[test]
    fn from_file_shifts_positional1_to_the_data_file() {
        let c = cli(Some("expr.jsonata"), Some("data.json"), None, false);
        assert_eq!(
            resolve(&c),
            Ok((
                ExpressionSource::File("expr.jsonata".into()),
                InputSource::File("data.json".into())
            ))
        );
    }

    #[test]
    fn from_file_with_no_positionals_reads_stdin() {
        let c = cli(Some("expr.jsonata"), None, None, false);
        assert_eq!(
            resolve(&c),
            Ok((
                ExpressionSource::File("expr.jsonata".into()),
                InputSource::Stdin
            ))
        );
    }

    #[test]
    fn from_file_with_two_positionals_is_an_error() {
        let c = cli(Some("expr.jsonata"), Some("extra1"), Some("extra2"), false);
        assert!(resolve(&c).is_err());
    }
}
